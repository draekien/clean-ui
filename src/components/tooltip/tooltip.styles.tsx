import { ThemeUIStyleObject } from 'theme-ui';
import { Position } from '../../hooks/useAutoPosition';

export const tooltipContainerCss: ThemeUIStyleObject = {
  display: 'inline-block',
  position: 'relative',
  width: '100%',
};

export const tooltipCss = (position: Position): ThemeUIStyleObject => {
  const css: any = {
    position: 'absolute',
    px: 'sm',
    py: 'xs',
    color: 'info-400',
    textAlign: 'center',
    backgroundColor: 'info-000',
    borderRadius: 'sm',
    variant: 'text.small',
    width: 'max-content',
    maxWidth: '20rem',
    visibility: 'hidden',
    zIndex: 'toast',
  };

  if (position === 'top') {
    css.marginBottom = '0.5rem';
    css.bottom = '100%';
    css.left = '50%';
    css.transform = 'translateX(-50%)';
    css.visibility = 'visible';
  }

  if (position === 'right') {
    css.marginLeft = '0.5rem';
    css.left = '100%';
    css.top = '50%';
    css.transform = 'translateY(-50%)';
    css.visibility = 'visible';
  }

  if (position === 'bottom') {
    css.marginTop = '0.5rem';
    css.top = '100%';
    css.left = '50%';
    css.transform = 'translateX(-50%)';
    css.visibility = 'visible';
  }

  if (position === 'left') {
    css.marginRight = '0.5rem';
    css.right = '100%';
    css.top = '50%';
    css.transform = 'translateY(-50%)';
    css.visibility = 'visible';
  }

  return css;
};

export const tooltipArrowCss = (position: Position): ThemeUIStyleObject => {
  const css: any = {
    position: 'absolute',
    width: 0,
    height: 0,
    borderColor: 'transparent',
    borderStyle: 'solid',
    zIndex: 'toast',
  };

  if (position === 'top') {
    css.left = '50%';
    css.bottom = '-0.25rem';
    css.marginLeft = '-0.25rem';
    css.borderWidth = '0.25rem 0.25rem 0';
    css.borderTopColor = 'info-000';
  }

  if (position === 'right') {
    css.top = '50%';
    css.right = '100%';
    css.marginTop = '-0.25rem';
    css.borderWidth = '0.25rem 0.25rem 0.25rem 0';
    css.borderRightColor = 'info-000';
  }

  if (position === 'bottom') {
    css.left = '50%';
    css.top = '-0.25rem';
    css.marginLeft = '-0.25rem';
    css.borderWidth = '0 0.25rem 0.25rem';
    css.borderBottomColor = 'info-000';
  }

  if (position === 'left') {
    css.top = '50%';
    css.left = '100%';
    css.marginTop = '-0.25rem';
    css.borderWidth = '0.25rem 0 0.25rem 0.25rem';
    css.borderLeftColor = 'info-000';
  }

  return css;
};
