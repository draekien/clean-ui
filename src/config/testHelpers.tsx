import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
