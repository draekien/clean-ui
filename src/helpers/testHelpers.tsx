import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
