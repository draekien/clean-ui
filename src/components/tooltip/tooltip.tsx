/** @jsxImportSource theme-ui */
import React, { createRef, useEffect, useState } from 'react';
import { Direction, Position, useAutoPosition } from '../../hooks/useAutoPosition';
import * as styles from './tooltip.styles';

const MARGIN = 12;
const DELAY = 300;

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
  text: React.ReactNode;
  children: React.ReactNode;
  direction?: Direction;
}

export const Tooltip: React.FC<TooltipProps> = ({
  text,
  children,
  direction = 'auto',
}) => {
  const fixedDirection = direction !== 'auto' ? direction : null;
  const [tooltipPosition, setTooltipPosition] = useState<Position>(fixedDirection);
  const [visible, setVisible] = useState(false);

  const element = createRef<HTMLDivElement>();
  const directionalElement = createRef<HTMLDivElement>();

  const position = useAutoPosition({
    direction,
    element,
    directionalElement,
    margin: MARGIN,
    visible,
  });

  useEffect(() => {
    setTooltipPosition(position);
  }, [position, visible, direction]);

  let timer: ReturnType<typeof setTimeout>;

  const handleOnFocus = () => {
    timer = setTimeout(() => {
      setVisible(true);
    }, DELAY);
  };

  const handleOnBlur = () => {
    clearTimeout(timer);
    setVisible(false);

    !fixedDirection && setTooltipPosition(null);
  };

  const handleOnMouseEnter = () => {
    timer = setTimeout(() => {
      setVisible(true);
    }, DELAY);
  };

  const handleOnMouseLeave = () => {
    clearTimeout(timer);
    setVisible(false);

    !fixedDirection && setTooltipPosition(null);
  };

  const handleOnMouseDown = () => {
    if (visible) {
      clearTimeout(timer);
      setVisible(false);

      !fixedDirection && setTooltipPosition(null);
      return;
    }

    setVisible(true);
  };

  return (
    <div
      ref={element}
      sx={styles.tooltipContainerCss}
      onFocus={handleOnFocus}
      onBlur={handleOnBlur}
      onMouseEnter={handleOnMouseEnter}
      onMouseLeave={handleOnMouseLeave}
      onMouseDown={handleOnMouseDown}>
      {visible && (
        <div ref={directionalElement} sx={styles.tooltipCss(tooltipPosition)}>
          <div sx={styles.tooltipArrowCss(tooltipPosition)} />
          {text}
        </div>
      )}
      {children}
    </div>
  );
};
