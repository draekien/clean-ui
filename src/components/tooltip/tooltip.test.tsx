import React from 'react';
import { themedRender } from '../../helpers/testHelpers';
import { Tooltip } from './tooltip';

const Element = () => <div id="test" style={{ width: '100px', height: '100px' }} />;

describe('Tooltip Component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('Should match snapshot', () => {
    const { container } = themedRender(
      <Tooltip text="tooltip">
        <Element />
      </Tooltip>
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('div').length).toEqual(2);
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
});
