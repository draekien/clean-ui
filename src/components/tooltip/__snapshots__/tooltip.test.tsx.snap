// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Component Should match snapshot 1`] = `
<div>
  <div
    class="css-108m3ho-Tooltip"
  >
    <div
      id="test"
      style="width: 100px; height: 100px;"
    />
  </div>
</div>
`;
